docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+bitmap. Implementations may choose to always allocate one anyway as a matter of
+convenience, but this should be noted when memory is being shared.
+
+Nested type arrays have their own null bitmap and null count regardless of
+the null count and null bits of their child arrays.
+
+Primitive
+---------
+
+A primitive value array represents a fixed-length array of values each having
+the same physical slot width typically measured in bytes, though the spec also
+provides for bit-packed types (e.g. boolean values encoded in bits).
+
+Internally, the array contains a contiguous memory buffer whose total size is
+equal to the slot width multiplied by the array length. For bit-packed types,
+the size is rounded up to the nearest byte.
+
+The associated null bitmap is contiguously allocated (as described above) but
+does not need to be adjacent in memory to the values buffer.
+
+**Example Layout: Int32 Array**
+
+For example a primitive array of int32s: ::
+
+    [1, null, 2, 4, 8]
+
+Would look like: ::
+
+    * Length: 5, Null count: 1
+    * Null bitmap buffer:
+
+      |Byte 0 (validity bitmap) | Bytes 1-63            |
+      |-------------------------|-----------------------|
+      | 00011101                | 0 (padding)           |
+
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-15 | Bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | unspecified | 2           | 4           | 8           | unspecified |
+
+**Example Layout: Non-null int32 Array**
+
+``[1, 2, 3, 4, 8]`` has two possible layouts: ::
+
+    * Length: 5, Null count: 0
+    * Null bitmap buffer:
+
+      | Byte 0 (validity bitmap) | Bytes 1-63            |
+      |--------------------------|-----------------------|
+      | 00011111                 | 0 (padding)           |
+
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | bytes 12-15 | bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | 2           | 3           | 4           | 8           | unspecified |
+
+or with the bitmap elided: ::
+
+    * Length 5, Null count: 0
+    * Null bitmap buffer: Not required
+    * Value Buffer:
+
+      |Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | bytes 12-15 | bytes 16-19 | Bytes 20-63 |
+      |------------|-------------|-------------|-------------|-------------|-------------|
+      | 1          | 2           | 3           | 4           | 8           | unspecified |
+
+Variable-size
+-------------
+
+Each value in this layout type consists of 0 or more bytes. While
+primitive arrays have a single values buffer, variable-size binary
+have an **offsets** buffer and **data** buffer
+
+The offsets buffer contains `length + 1` signed integers (either
+32-bit or 64-bit), which encode the start position of each element in
```

Comment:
"either 32-bit or 64-bit, depending on the type" perhaps? (to not give the illusion that offsets themselves are variable-sized)